File: samples/Windows/java/TableManagerSamples/OpenPositionNetting/src/TableListener.java
package openpositionnetting;

import java.util.List;

import com.fxcore2.*;
import common.*;

public class TableListener implements IO2GTableListener {
    private OrderMonitorNetting mOrderMonitor;
    private ResponseListener mResponseListener;
    private String mRequestID;
    private O2GTradesTable mTradesTable;

    // ctor
    public TableListener(ResponseListener responseListener) {
        mOrderMonitor = null;
        mResponseListener = responseListener;
        mRequestID = "";
        mTradesTable = null;
    }

    public void setRequestID(String sRequestID) {
        mRequestID = sRequestID;
    }

    public void setTradesTable(O2GTradesTable tradesTable) {
        mTradesTable = tradesTable;
    }

    public void onAdded(String sRowID, O2GRow rowData) {
        O2GTableType type = rowData.getTableType();
        switch (type) {
        case ORDERS:
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mRequestID.equals(orderRow.getRequestID())) {
                if ((OrderMonitorNetting.isClosingOrder(orderRow) || OrderMonitorNetting.isOpeningOrder(orderRow)) &&
                        mOrderMonitor == null) {
                    System.out.println(String.format("The order has been added. Order ID: %s, Rate: %s, Time In Force: %s",
                            orderRow.getOrderID(),
                            orderRow.getRate(),
                            orderRow.getTimeInForce()));
                    O2GTradeRow trade = null;
                    String sTradeID = orderRow.getTradeID();
                    if (mTradesTable != null) {
                        for (int j = 0; j < mTradesTable.size(); j++) {
                            if (sTradeID.equals(mTradesTable.getRow(j).getTradeID())) {
                                trade = mTradesTable.getRow(j);
                                break;
                            }
                        }
                    }
                    if (trade == null) {
                        mOrderMonitor = new OrderMonitorNetting(orderRow);
                    } else {
                        mOrderMonitor = new OrderMonitorNetting(orderRow, trade.getAmount());
                    }
                }
            }
            break;
        case TRADES:
            O2GTradeRow tradeRow = (O2GTradeRow)rowData;
            if (mOrderMonitor != null) {
                mOrderMonitor.onTradeAdded(tradeRow);
                if (mOrderMonitor.isOrderCompleted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        case CLOSED_TRADES:
            O2GClosedTradeRow closedTradeRow = (O2GClosedTradeRow)rowData;
            if (mOrderMonitor != null) {
                mOrderMonitor.onClosedTradeAdded(closedTradeRow);
                if (mOrderMonitor.isOrderCompleted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        case MESSAGES:
            O2GMessageRow messageRow = (O2GMessageRow)rowData;
            if (mOrderMonitor != null) {
                mOrderMonitor.onMessageAdded(messageRow);
                if (mOrderMonitor.isOrderCompleted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
            break;
        }
    }

    public void onChanged(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ACCOUNTS) {
            O2GAccountTableRow account = (O2GAccountTableRow)rowData;
            System.out.println(String.format("Balance: %.2f, Equity: %.2f", account.getBalance(), account.getEquity()));
        } else if (rowData.getTableType() == O2GTableType.TRADES) {
            if (mOrderMonitor != null) {
                mOrderMonitor.onTradeUpdated((O2GTradeRow)rowData);
                if (mOrderMonitor.isOrderCompleted()) {
                    printResult();
                    mResponseListener.stopWaiting();
                }
            }
        }
    }

    public void onDeleted(String sRowID, O2GRow rowData) {
        if (rowData.getTableType() == O2GTableType.ORDERS) {
            O2GOrderRow orderRow = (O2GOrderRow)rowData;
            if (mRequestID.equals(orderRow.getRequestID())) {
                System.out.println(String.format("The order has been deleted. Order ID: %s", orderRow.getOrderID()));
                mOrderMonitor.onOrderDeleted(orderRow);
                if (mOrderMonitor != null) {
                    if (mOrderMonitor.isOrderCompleted()) {
                        printResult();
                        mResponseListener.stopWaiting();
                    }
                }
            }
        }
    }

    public void onStatusChanged(O2GTableStatus status) {
    }

    private void printResult() {
        if (mOrderMonitor != null) {
            OrderMonitorNetting.ExecutionResult result = mOrderMonitor.getResult();
            List<O2GTradeRow> trades;
            List<O2GTradeRow> updatedTrades;
            List<O2GClosedTradeRow> closedTrades;
            O2GOrderRow order = mOrderMonitor.getOrder();
            String sOrderID = order.getOrderID();
            trades = mOrderMonitor.getTrades();
            updatedTrades = mOrderMonitor.getUpdatedTrades();
            closedTrades = mOrderMonitor.getClosedTrades();

            switch (result) {
            case Canceled:
                if (trades.size() > 0) {
                    printTrades(trades, sOrderID);
                    printClosedTrades(closedTrades, sOrderID);
                    System.out.println(String.format("A part of the order has been canceled. Amount = %s", mOrderMonitor.getRejectAmount()));
                } else {
                    System.out.println(String.format("The order: OrderID = %s  has been canceled.", sOrderID));
                    System.out.println(String.format("The cancel amount = %s.", mOrderMonitor.getRejectAmount()));
                }
                break;
            case FullyRejected:
                System.out.println(String.format("The order has been rejected. OrderID = %s", sOrderID));
                System.out.println(String.format("The rejected amount = %s", mOrderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s", mOrderMonitor.getRejectMessage()));
                break;
            case PartialRejected:
                printTrades(trades, sOrderID);
                printUpdatedTrades(updatedTrades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                System.out.println(String.format("A part of the order has been rejected. Amount = %s", mOrderMonitor.getRejectAmount()));
                System.out.println(String.format("Rejection cause: %s ", mOrderMonitor.getRejectMessage()));
                break;
            case Executed:
                printTrades(trades, sOrderID);
                printUpdatedTrades(updatedTrades, sOrderID);
                printClosedTrades(closedTrades, sOrderID);
                break;
            }
        }
    }

    private void printTrades(List<O2GTradeRow> trades, String sOrderID) {
        if (trades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been opened:", sOrderID));

        for (int i = 0; i < trades.size(); i++) {
            O2GTradeRow trade = trades.get(i);
            String sTradeID = trade.getTradeID();
            int iAmount = trade.getAmount();
            double dRate = trade.getOpenRate();
            System.out.println(String.format("Trade ID: %s; Amount: %s; Rate: %s", sTradeID, iAmount, dRate));
        }
    }

    private void printUpdatedTrades(List<O2GTradeRow> updatedTrades, String sOrderID) {
        if (updatedTrades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been updated:", sOrderID));

        for (int i = 0; i < updatedTrades.size(); i++) {
            O2GTradeRow trade = updatedTrades.get(i);
            String sTradeID = trade.getTradeID();
            int iAmount = trade.getAmount();
            double dRate = trade.getOpenRate();
            System.out.println(String.format("Trade ID: %s; Amount: %s; Rate: %s", sTradeID, iAmount, dRate));
        }
    }

    private void printClosedTrades(List<O2GClosedTradeRow> closedTrades, String sOrderID) {
        if (closedTrades.size() == 0)
            return;
        System.out.println(String.format("For the order: OrderID = %s the following positions have been closed: ", sOrderID));

        for (int i = 0; i < closedTrades.size(); i++) {
            O2GClosedTradeRow closedTrade = closedTrades.get(i);
            String sTradeID = closedTrade.getTradeID();
            int iAmount = closedTrade.getAmount();
            double dRate = closedTrade.getCloseRate();
            System.out.println(String.format("Closed Trade ID: %s; Amount: %s; Closed Rate: %s", sTradeID, iAmount, dRate));
        }
    }

    public void subscribeEvents(O2GTableManager manager) {
        O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.ACCOUNTS);
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.TRADES);
        O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.MESSAGES);
        O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.CLOSED_TRADES);
        accountsTable.subscribeUpdate(O2GTableUpdateType.UPDATE, this);
        ordersTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.subscribeUpdate(O2GTableUpdateType.DELETE, this);
        tradesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        tradesTable.subscribeUpdate(O2GTableUpdateType.UPDATE, this);
        closedTradesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
        messagesTable.subscribeUpdate(O2GTableUpdateType.INSERT, this);
    }

    public void unsubscribeEvents(O2GTableManager manager) {
        O2GAccountsTable accountsTable = (O2GAccountsTable)manager.getTable(O2GTableType.ACCOUNTS);
        O2GOrdersTable ordersTable = (O2GOrdersTable)manager.getTable(O2GTableType.ORDERS);
        O2GTradesTable tradesTable = (O2GTradesTable)manager.getTable(O2GTableType.TRADES);
        O2GMessagesTable messagesTable = (O2GMessagesTable)manager.getTable(O2GTableType.MESSAGES);
        O2GClosedTradesTable closedTradesTable = (O2GClosedTradesTable)manager.getTable(O2GTableType.CLOSED_TRADES);
        accountsTable.unsubscribeUpdate(O2GTableUpdateType.UPDATE, this);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        ordersTable.unsubscribeUpdate(O2GTableUpdateType.DELETE, this);
        tradesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        tradesTable.unsubscribeUpdate(O2GTableUpdateType.UPDATE, this);
        closedTradesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
        messagesTable.unsubscribeUpdate(O2GTableUpdateType.INSERT, this);
    }
}

File: samples/Windows/cpp/TableManagerSamples/OpenPosition/source/TableListener.cpp
#include "stdafx.h"
#include "OrderMonitor.h"
#include "ResponseListener.h"
#include "TableListener.h"

TableListener::TableListener(ResponseListener *responseListener)
{
    mRefCount = 1;
    mOrderMonitor = NULL;
    mResponseListener = responseListener;
    mResponseListener->addRef();
    mRequestID = "";
    std::cout.precision(2);
}

TableListener::~TableListener(void)
{
    if (mResponseListener)
        mResponseListener->release();
}

long TableListener::addRef()
{
    return InterlockedIncrement(&mRefCount);
}

long TableListener::release()
{
    InterlockedDecrement(&mRefCount);
    if(mRefCount == 0)
        delete this;
    return mRefCount;
}

/** Set request. */
void TableListener::setRequestID(const char *sRequestID)
{
    mRequestID = sRequestID;
}

void TableListener::onAdded(const char *rowID, IO2GRow *row)
{

    O2GTable type = row->getTableType();
    switch(type)
    {
    case Orders:
    {
        IO2GOrderRow *order = static_cast<IO2GOrderRow *>(row);
        if( mRequestID == order->getRequestID() && 
            (isClosingOrder(order) || isOpeningOrder(order)))
        {
            std::cout << "The order has been added. OrderID='" << order->getOrderID() << "', "
                    << "Rate='" << order->getRate() << "', "
                    << "TimeInForce='" << order->getTimeInForce() << "'"
                    << std::endl;
            mOrderMonitor = new OrderMonitor(order);
        }
    }
    break;
    case Trades:
    {
        IO2GTradeRow *trade = static_cast<IO2GTradeRow *>(row);
        if (mOrderMonitor)
        {
            mOrderMonitor->onTradeAdded(trade);
            if (mOrderMonitor->isOrderCompleted())
            {
                printResult();
                mResponseListener->stopWaiting();
            }
        }
    }
    break;
    case ClosedTrades:
    {
        IO2GClosedTradeRow *closedTrade = static_cast<IO2GClosedTradeRow *>(row);
        if (mOrderMonitor)
        {
            mOrderMonitor->onClosedTradeAdded(closedTrade);
            if (mOrderMonitor->isOrderCompleted())
            {
                printResult();
                mResponseListener->stopWaiting();
            }
        }
    }
    break;
    case Messages:
    {
        IO2GMessageRow *message = static_cast<IO2GMessageRow *>(row);
        if (mOrderMonitor)
        {
            mOrderMonitor->onMessageAdded(message);
            if (mOrderMonitor->isOrderCompleted())
            {
                printResult();
                mResponseListener->stopWaiting();
            }
        }
    }
    break;
    }

}

void TableListener::onChanged(const char *rowID, IO2GRow *row)
{
    O2GTable type = row->getTableType();
    if (type == Accounts)
    {
        IO2GAccountTableRow *account = (IO2GAccountTableRow *)row;
        std::cout << "The balance has been changed. AccountID=" << account->getAccountID() << ", "
                << "Balance=" << std::fixed << account->getBalance() << ", "
                << "Equity=" << std::fixed << account->getEquity()
                << std::endl;
    }
}

void TableListener::onDeleted(const char *rowID, IO2GRow *row)
{
    IO2GOrderRow *order = static_cast<IO2GOrderRow *>(row);
    if( mRequestID == order->getRequestID() )
    {
        std::cout << "The order has been deleted. OrderID='" << order->getOrderID() << "'"
                << std::endl;
        mOrderMonitor->onOrderDeleted(order);
        if (mOrderMonitor)
        {
            if (mOrderMonitor->isOrderCompleted())
            {
                printResult();
                mResponseListener->stopWaiting();
            }
        }
    }
}

void TableListener::onStatusChanged(O2GTableStatus status)
{
}

void TableListener::printResult()
{
    if (mOrderMonitor)
    {
        OrderMonitor::ExecutionResult result = mOrderMonitor->getResult();
        std::vector<IO2GTradeRow*> trades;
        std::vector<IO2GClosedTradeRow*> closedTrades;
        O2G2Ptr<IO2GOrderRow> order = mOrderMonitor->getOrder();
        std::string orderID = order->getOrderID();
        mOrderMonitor->getTrades(trades);
        mOrderMonitor->getClosedTrades(closedTrades);

        switch (result)
        {
        case OrderMonitor::Canceled:
        {
            if (trades.size() > 0)
            {
                printTrades(trades, orderID);
                printClosedTrades(closedTrades, orderID);
                std::cout << "A part of the order has been canceled. "
                        << "Amount = " << mOrderMonitor->getRejectAmount() << std::endl;
            }
            else
            {
                std::cout << "The order: OrderID = " << orderID << " has been canceled" << std::endl;
                std::cout << "The cancel amount = " << mOrderMonitor->getRejectAmount() << std::endl;
            }
        }
        break;
        case OrderMonitor::FullyRejected:
        {
            std::cout << "The order has been rejected. OrderID = " << orderID << std::endl;
            std::cout << "The rejected amount = " << mOrderMonitor->getRejectAmount() << std::endl;;
            std::cout << "Rejection cause: " << mOrderMonitor->getRejectMessage() << std::endl;
        }
        break;
        case OrderMonitor::PartialRejected:
        {
            printTrades(trades, orderID);
            printClosedTrades(closedTrades, orderID);
            std::cout << "A part of the order has been rejected. "
                    << "Amount = " << mOrderMonitor->getRejectAmount() << std::endl;
            std::cout << "Rejection cause: " << mOrderMonitor->getRejectMessage() << std::endl;
        }
        break;
        case OrderMonitor::Executed:
        {
            printTrades(trades, orderID);
            printClosedTrades(closedTrades, orderID);
        }
        break;
        }
    }
}

void TableListener::printTrades(std::vector<IO2GTradeRow*> &trades, std::string &sOrderID)
{
    if (trades.size() == 0)
        return;
    std::cout << "For the order: OrderID=" << sOrderID << " the following positions have been opened: " << std::endl;
    for (size_t i = 0; i < trades.size(); ++i)
    {
        O2G2Ptr<IO2GTradeRow> trade = trades[i];
        std::string tradeID = trade->getTradeID();
        int amount = trade->getAmount();
        double rate = trade->getOpenRate();
        std::cout << "Trade ID: " << tradeID << ", "
                << "Amount: " << amount << ", "
                << "Rate: " << rate << std::endl;
    }
}

void TableListener::printClosedTrades(std::vector<IO2GClosedTradeRow*> &closedTrades, std::string &sOrderID)
{
    if (closedTrades.size() == 0)
        return;
    std::cout << "For the order: OrderID=" << sOrderID << " the following positions have been closed: " << std::endl;
    for (size_t i = 0; i < closedTrades.size(); ++i)
    {
        IO2GClosedTradeRow *closedTrade = closedTrades[i];
        std::string tradeID = closedTrade->getTradeID();
        int amount = closedTrade->getAmount();
        double rate = closedTrade->getCloseRate();
        std::cout << "Closed Trade ID: " << tradeID << ", "
                << "Amount: " << amount << ", "
                << "Closed Rate: " << rate << std::endl;
        closedTrade->release();
    }
}

void TableListener::subscribeEvents(IO2GTableManager *manager)
{
    O2G2Ptr<IO2GAccountsTable> accountsTable = (IO2GAccountsTable *)manager->getTable(Accounts);
    O2G2Ptr<IO2GOrdersTable> ordersTable = (IO2GOrdersTable *)manager->getTable(Orders);
    O2G2Ptr<IO2GTradesTable> tradesTable = (IO2GTradesTable *)manager->getTable(Trades);
    O2G2Ptr<IO2GMessagesTable> messagesTable = (IO2GMessagesTable *)manager->getTable(Messages);
    O2G2Ptr<IO2GClosedTradesTable> closedTradesTable = (IO2GClosedTradesTable *)manager->getTable(ClosedTrades);

    accountsTable->subscribeUpdate(Update, this);
    ordersTable->subscribeUpdate(Insert, this);
    ordersTable->subscribeUpdate(Delete, this);
    tradesTable->subscribeUpdate(Insert, this);
    tradesTable->subscribeUpdate(Update, this);
    closedTradesTable->subscribeUpdate(Insert, this);
    messagesTable->subscribeUpdate(Insert, this);
}

void TableListener::unsubscribeEvents(IO2GTableManager *manager)
{
    O2G2Ptr<IO2GAccountsTable> accountsTable = (IO2GAccountsTable *)manager->getTable(Accounts);
    O2G2Ptr<IO2GOrdersTable> ordersTable = (IO2GOrdersTable *)manager->getTable(Orders);
    O2G2Ptr<IO2GTradesTable> tradesTable = (IO2GTradesTable *)manager->getTable(Trades);
    O2G2Ptr<IO2GMessagesTable> messagesTable = (IO2GMessagesTable *)manager->getTable(Messages);
    O2G2Ptr<IO2GClosedTradesTable> closedTradesTable = (IO2GClosedTradesTable *)manager->getTable(ClosedTrades);

    accountsTable->unsubscribeUpdate(Update, this);
    ordersTable->unsubscribeUpdate(Insert, this);
    ordersTable->unsubscribeUpdate(Delete, this);
    tradesTable->unsubscribeUpdate(Insert, this);
    tradesTable->unsubscribeUpdate(Update, this);
    closedTradesTable->unsubscribeUpdate(Insert, this);
    messagesTable->unsubscribeUpdate(Insert, this);
}
